### PR TITLE
Add Macro.classify_atom/1 and Macro.inspect_atom/2

### PR DESCRIPTION
**Pull request by josevalim**
Today the formatter still uses some private APIs
and this commit aims to expose some of them.
In particular, Macro.classify_atom/1 was added to
expose if an atom is an alias, an identifier, quoted,
or unquoted.

Note, however, there is no API for operators yet.
That's because operators can fall within three
distinct areas:

  * unquoted and callable, such as, `+`, `-`, and
    most operators

  * unquoted and not callable, such as, `.`, `..`,
    and `..//`, which would be ambiguous when used
    as `Module.OP()`

  * quoted but callable, such as `::`, which is
    ambiguous in the atom syntax but not as calls

This information still remains private, especially
because most times operators are used, we often still
want to learn about their precedence too.

We also exposed `inspect_atom/2` with three
distinct clauses to cover the different scenarios
atoms can be inspected as.

### CODE REVIEW

File: lib/elixir/lib/atom.ex [hunk 1]
```diff
@@ -75,6 +75,247 @@ defmodule Atom do
...
+
+    * `:alias` - the atom represents an alias
+
+    * `:identifier` - the atom can be used as a variable or local function call
+
+    * `:unquoted` - the atom can be used in its unquoted form,
+      includes operators and atoms with `@` in them
+
+    * `:quoted` - all other atoms which can only be used in their quoted form
+
```
Comment (josevalim) [inline lib/elixir/lib/atom.ex:91]:
```suggestion

  Note operators are going to either be unquoted, such as `:+` and
  most operators, or quoted, such as `:"::"`.
```

**Comment by ericmj** [comment]
I feel like `Atom.classify/1` fits better as `Macro.classify_atom/1`, because in the Atom module atoms are just atoms. But this function is about how we represent atoms in code and not about atoms in general. Same with `Atom.inspect/1`, isn't it a subset of `Macro.to_string/1`?

**Comment by josevalim** [comment]
@ericmj agreed. I think we should have `classify_atom` and `inspect_atom` and for `inspect_atom` we have the first key be: `:literal | :key | :remote`.

**Comment by josevalim** [comment]
@ericmj please check :)

**Review by ericmj — APPROVED** [review]
💜

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**